src/qibolab/instruments/qblox/controller.py

```diff
@@ -493,48 +490,52 @@ def _sweep_recursion(
                         for sft_iteration in range(num_full_sft_iterations + 1):
                             _nshots = min(
                                 max_rt_nshots, nshots - sft_iteration * max_rt_nshots
                             )
-                            self._sweep_recursion(
+
+                            res = self._execute_pulse_sequence(
                                 qubits,
                                 sequence,
-                                options,
-                                *sweepers,
-                                results=results,
+                                replace(options, nshots=_nshots),
+                                sweepers,
                             )
+                            result_chunks.append(res)
+                        result = self._combine_result_chunks(result_chunks)
+                        self._add_to_results(sequence, results, result)
                     else:
-                        for _ in range(nshots):
-                            num_bins = 1
-                            for sweeper in sweepers[1:]:
-                                num_bins *= len(sweeper.values)
-                            sweeper = sweepers[0]
-                            max_rt_iterations = (SEQUENCER_MEMORY) // num_bins
-                            num_full_sft_iterations = (
-                                len(sweeper.values) // max_rt_iterations
-                            )
-                            num_bins = nshots * max_rt_iterations
-                            for sft_iteration in range(num_full_sft_iterations + 1):
-                                _from = sft_iteration * max_rt_iterations
-                                _to = min(
-                                    (sft_iteration + 1) * max_rt_iterations,
-                                    len(sweeper.values),
-                                )
-                                _values = sweeper.values[_from:_to]
-                                split_sweeper = Sweeper(
-                                    parameter=sweeper.parameter,
-                                    values=_values,
-                                    pulses=sweeper.pulses,
-                                    qubits=sweeper.qubits,
-                                )
+                        raise ValueError(
+                            f"Requested sweep has {sweepers_repetitions} total number of sweep points. "
+                            f"Maximum supported is {SEQUENCER_MEMORY}"
+                        )
 
-                                self._sweep_recursion(
-                                    qubits,
-                                    sequence,
-                                    options,
-                                    *((split_sweeper,) + sweepers[1:]),
-                                    results=results,
-                                )
+    @staticmethod
+    def _combine_result_chunks(result_chunks):
```

Comment:
I am thinking of implementing this functionality as some sort of `append` method for data types `IntegratedResults` and `SampleResults`, instead of doing the gymnastics here. The existing methods of those classes make them look like they were designed to be immutable. Is this impression true? If no, I guess just a simple append method that changes each object in place is enough. If yes, I can implement classmethods that accept multiple instances and create a new instance containing the appended (across a certain axis) data. Any comments?

Reply:
> The existing methods of those classes make them look like they were designed to be immutable. Is this impression true?

It seems that these objects contain numpy arrays which are mutable, so I am not sure if it is possible to make them immutable. Since they are part of the outer interface, it could be good to make them immutable, so when the user executes something they get back an object that doesn't change, however personally I don't have strong opinion on that. But it seems that it is not possible due to the array anyway, so I do not see any issue with your proposal.

Reply:
yeah, they are for sure not immutable objects. I was more thinking along the lines "they are not immutable, but it seems they are intended to be treated as immutable, or used in a manner that pretends as if they are immutable". I think it would be nicer to make them frozen dataclasses. I would do it as another PR. I am not sure if any party currently is relying on the fact that they are mutable. For now I can just go with the classmethod/factory method approach, so that the classes remain immutable-friendly.

Reply:
To the best of my understanding, they already have an append-like function, i.e.
https://github.com/qiboteam/qibolab/blob/d080141c9a4fd2160e5b159205eded3256333c75/src/qibolab/result.py#L20-L21
and, consequently, `__iadd__` is automatically defined as:
```py
self = self + other
```

So, if you want to append, you could use the `+=` operator.

Reply:
As @stavros11 said, it could be a good idea to make them immutable (despite I would avoid making the wrestling required to mimic immutability with NumPy array, which are intrinsically mutable, and just pretend them to be immutable, avoiding mutations in Qibolab).

In any case, if you'd like to implement such a feature, please target 0.2. It should definitely be grouped together with the other breaking (and potentially-breaking) changes.

Reply:
Btw, my original proposal was to avoid making classes for these results altogether, since they are just the downloaded arrays.
If they are composite, they should have been composite dtypes (see rec arrays).

We could have labels, but they would be essentially the execution parameters. And if we could stick to bare arrays, we could also avoid:
https://github.com/qiboteam/qibolab/blob/d080141c9a4fd2160e5b159205eded3256333c75/src/qibolab/execution_parameters.py#L40-L51

In any case, they are array wrappers. I'd try to keep them as simple as possible...